### PR TITLE
Improve Consumer with multiple subjects

### PR DESCRIPTION
**Pull request by Jarema**
While adding more complex test scenarios, some issues in multiple-filters approach surfaced.
This PR adds those tests and addresses the issues.

Signed-off-by: Tomasz Pietrek <tomasz@nats.io>

### CODE REVIEW

File: server/consumer.go [hunk 1]
```diff
@@ -3313,31 +3310,36 @@ func (o *consumer) getNextMsg() (*jsPubMsg, uint64, error) {
 	// to avoid reflection.
 	if o.subjf != nil && len(o.subjf) > 1 {
 		sort.Slice(o.subjf, func(i, j int) bool {
+			if o.subjf[j].pmsg != nil && o.subjf[i].pmsg == nil {
```
Comment (Jarema) [inline server/consumer.go:3313]: Checking for `pmsg` not being nil mainly prevents errors when two equal filter sequences (at the end of the stream) happen, and one has the last message.

Through all tests, it looks safe to skip checking if they are equal and just order them properly based `pmsg` absence or not.

File: server/consumer.go [hunk 2]
```diff
@@ -3282,20 +3281,18 @@ func (o *consumer) getNextMsg() (*jsPubMsg, uint64, error) {
 			o.mu.Unlock()
 			pmsg := getJSPubMsgFromPool()
 			sm, sseq, err := store.LoadNextMsg(filterSubject, filterWC, nextSeq, &pmsg.StoreMsg)
-			o.mu.Lock()
 
 			filter.err = err
```
Comment (derekcollison) [inline server/consumer.go:3285]: Can these be manipulated safely without the consumer lock?

File: server/consumer.go [hunk 3]
```diff
@@ -3282,20 +3281,18 @@ func (o *consumer) getNextMsg() (*jsPubMsg, uint64, error) {
...
 
 			if sm != nil {
 				filter.pmsg = pmsg
```
Comment (derekcollison) [inline server/consumer.go:3288]: Same here.

File: server/consumer.go [hunk 4]
```diff
@@ -3313,31 +3310,36 @@ func (o *consumer) getNextMsg() (*jsPubMsg, uint64, error) {
...
-			filter.currentSeq = filter.nextSeq
-			o.sseq = filter.currentSeq
-			returned := filter.pmsg
-			filter.pmsg = nil
-			return returned, 1, filter.err
-		}
-		if filter.err != nil {
-			lastErr = filter.err
-		}
+	filter := o.subjf[0]
```
Comment (derekcollison) [inline server/consumer.go:3324]: Do we know len(o.subjf) >= 1 here?

File: server/consumer.go (hunk 2 above)
Comment (Jarema) [reply server/consumer.go]: During normal consumer operations, yes.
Added a test for that.

However... I think it's a good point and it's safer to move it before manipulating filter fields - especially since this could surface bugs after some changes to consumers, their updates etc. 

Thx!

File: server/consumer.go (hunk 4 above)
Comment (Jarema) [reply server/consumer.go:3329]: It should always be nil if there are no filters.
Your question gave me an idea when it could be not (when you remove all filters, and I protected that case and added test.

However, I also added a check for an empty slice to ensure it doesn't happen in some edge cases.
Thanks for catching this!

File: server/consumer.go (hunk 2 above)
Comment (derekcollison) [reply server/consumer.go]: Why are you unlocking and locking in the first place? To not hold the lock during load?

Comment (Jarema) [reply server/consumer.go]: The locking and unlocking in consumer was added here: https://github.com/nats-io/nats-server/pull/3922/files, specifically in commit https://github.com/nats-io/nats-server/pull/3922/commits/13167f46b9884f3bb4b647a49f928c02af6fb325 as a performance improvement under heavy load.

File: server/consumer.go [hunk 5]
```diff
@@ -3247,7 +3252,7 @@ func (o *consumer) getNextMsg() (*jsPubMsg, uint64, error) {
 	store := o.mset.store
 
 	// If no filters are specified, optimize to fetch just non-filtered messages.
-	if o.subjf == nil {
+	if o.subjf == nil || len(o.subjf) == 0 {
```
Comment (derekcollison) [inline server/consumer.go:3255]: Just do len() since len(nil) is valid.

File: server/consumer.go [hunk 6]
```diff
@@ -3313,31 +3315,36 @@ func (o *consumer) getNextMsg() (*jsPubMsg, uint64, error) {
 	// to avoid reflection.
 	if o.subjf != nil && len(o.subjf) > 1 {
```
Comment (derekcollison) [inline server/consumer.go:3316]: Same here, just do len.

**Review by derekcollison — APPROVED** [review]
LGTM
